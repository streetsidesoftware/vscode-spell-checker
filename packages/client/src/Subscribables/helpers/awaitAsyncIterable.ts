export async function awaitAsyncIterable<T>(i: AsyncIterable<T> | AsyncIterableIterator<T>): Promise<T[]> {
    const buffer: T[] = [];

    for await (const v of i) {
        buffer.push(v);
    }

    return buffer;
}
