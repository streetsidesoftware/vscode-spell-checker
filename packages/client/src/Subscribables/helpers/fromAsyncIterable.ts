import { disposeOf } from 'utils-disposables';

import { squelch } from '../../util/errors.js';
import { createEmitter } from '../createFunctions.js';
import type { Subscribable, SubscribableEvent } from '../Subscribables.js';

export function fromAsyncIterable<T>(iter: AsyncIterableIterator<T> | AsyncIterable<T>): Subscribable<T> {
    const emitter = createEmitter<T>();
    let stop = false;
    const disposeEventListener = emitter.onEvent(handleEvents);

    function handleEvents(e: SubscribableEvent) {
        if (e.name === 'onStart') {
            setTimeout(emitValues);
        }
        if (e.name === 'onStop' || e.name === 'onDone') {
            stop = true;
        }
    }

    async function _emitValues() {
        try {
            for await (const val of iter) {
                if (stop) break;
                emitter.notify(val);
            }
        } finally {
            disposeOf(disposeEventListener);
            emitter.done();
        }
    }

    const cErrors = squelch('fromAsyncIterable.emitValues');

    function emitValues() {
        _emitValues().catch(cErrors);
    }

    return emitter;
}
