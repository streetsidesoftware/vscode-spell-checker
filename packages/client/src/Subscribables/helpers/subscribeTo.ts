import type { DisposableLike } from 'utils-disposables';

import type { SubscribableLike, SubscribeFn, SubscriberLike } from '../Subscribables';

export function subscribeTo<T>(source: SubscribableLike<T>, subscriber: SubscriberLike<T>): DisposableLike {
    const src: SubscribeFn<T> = typeof source === 'function' ? source : (s) => source.subscribe(s);
    return src(subscriber) ?? (() => undefined);
}
