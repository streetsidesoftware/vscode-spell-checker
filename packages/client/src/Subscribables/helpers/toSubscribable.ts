import { createSubscribable } from '../createFunctions';
import type { Subscribable, SubscribableLike } from '../Subscribables';

export function fromSubscribableLike<T>(subscribable: SubscribableLike<T>): Subscribable<T> {
    return typeof subscribable === 'function' ? createSubscribable(subscribable) : subscribable;
}
