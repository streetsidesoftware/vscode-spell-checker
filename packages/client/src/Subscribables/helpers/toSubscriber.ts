import type { Subscriber, SubscriberFn, SubscriberLike } from '../Subscribables';

export function toSubscriber<T>(subscriberLike: SubscriberLike<T>): Subscriber<T> {
    if (typeof subscriberLike !== 'function') return subscriberLike;
    return {
        notify: subscriberLike,
    };
}

export function toSubscriberFn<T>(subscriberLike: SubscriberLike<T>): SubscriberFn<T> {
    if (typeof subscriberLike === 'function') return subscriberLike;
    return (value) => subscriberLike.notify(value);
}
