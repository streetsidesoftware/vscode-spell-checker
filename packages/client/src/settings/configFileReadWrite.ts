import { CSpellPackageSettings, CSpellSettings } from '@cspell/cspell-types';
import { assign as assignJson, parse as parseJsonc, stringify as stringifyJsonc } from 'comment-json';
import { isErrnoException } from 'common-utils/index.js';
import * as fs from 'fs-extra';
import { Uri } from 'vscode';
import { Utils as UriUtils } from 'vscode-uri';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import { CSpellUserSettings } from '../client';
import { ConfigReaderWriter, ConfigUpdateFn, extractKeys } from './configReaderWriter';
export type { ConfigUpdateFn } from './configReaderWriter';

const SymbolFormat = Symbol('format');

type HandlerDef = { match: RegExp; handler: (uri: Uri) => ConfigFileReaderWriter };

const handlers: HandlerDef[] = [
    { match: /package\.json$/i, handler: (uri) => new ConfigFileReaderWriterPackage(uri) },
    { match: /\.jsonc?$/i, handler: (uri) => new ConfigFileReaderWriterJson(uri) },
    { match: /\.ya?ml$/i, handler: (uri) => new ConfigFileReaderWriterYaml(uri) },
];

const spacesJson = 4;
const spacesPackage = 2;

/**
 *
 * @param uri - uri of the configuration file.
 * @param updateFn - function to be called with the config data to be updated. It should only return the fields to be update.
 *  A fields with a value of `undefined` will be removed from the file.
 *
 * @returns resolves if successful.
 */
export function updateConfigFile(uri: Uri, updateFn: ConfigUpdateFn): Promise<void> {
    const rw = createConfigFileReaderWriter(uri);
    return rw._update(updateFn);
}

export function readConfigFile(uri: Uri, defaultValueIfNotFound: CSpellSettings): Promise<CSpellSettings>;
export function readConfigFile(uri: Uri, defaultValueIfNotFound?: CSpellSettings): Promise<CSpellSettings | undefined>;
export async function readConfigFile(uri: Uri, defaultValueIfNotFound?: CSpellSettings): Promise<CSpellSettings | undefined> {
    try {
        const rw = createConfigFileReaderWriter(uri);
        return await rw._read();
    } catch (e) {
        if (isErrnoException(e) && e.code === 'ENOENT') {
            return Promise.resolve(defaultValueIfNotFound);
        }
        return Promise.reject(e);
    }
}

export function writeConfigFile(uri: Uri, cfg: CSpellSettings): Promise<void> {
    const rw = createConfigFileReaderWriter(uri);
    return rw.write(cfg);
}

export function isHandled(uri: Uri): boolean {
    return !!matchHandler(uri);
}

export function createConfigFileReaderWriter(uri: Uri): ConfigFileReaderWriter {
    return mustMatchHandler(uri);
}

function mustMatchHandler(uri: Uri): ConfigFileReaderWriter {
    const h = matchHandler(uri);
    if (h) return h.handler(uri);
    throw new UnhandledFileType(uri);
}

function matchHandler(uri: Uri): HandlerDef | undefined {
    const u = uri.with({ fragment: '', query: '' });
    const s = u.toString();
    for (const h of handlers) {
        if (!h.match.test(s)) continue;
        return h;
    }
    return undefined;
}

const settingsFileTemplate: CSpellSettings = {
    version: '0.2',
    ignorePaths: [],
    dictionaryDefinitions: [],
    dictionaries: [],
    words: [],
    ignoreWords: [],
    import: [],
};

interface PackageWithCSpell {
    cspell?: CSpellPackageSettings;
}

function orDefault<T>(p: Promise<T>, defaultValue: T): Promise<T> {
    return p.catch((e) => {
        if (e.code !== 'ENOENT') throw e;
        return defaultValue;
    });
}

export class UnhandledFileType extends Error {
    constructor(uri: Uri) {
        super(`Unhandled file type: "${UriUtils.basename(uri)}"`);
    }
}

export class SysLikeError extends Error {
    constructor(msg?: string, readonly code?: string, readonly errno?: number) {
        super(msg);
    }
}

export function parseJson(content: string): CSpellSettings {
    const formatting = detectFormatting(content);
    return injectFormatting(parseJsonc(content), formatting);
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function stringifyJson(obj: Object, spaces?: string | number | undefined, keepComments = true): string {
    const formatting = retrieveFormatting(obj);
    spaces = formatting?.spaces || spaces || spacesJson;
    const newlineAtEndOfFile = formatting?.newlineAtEndOfFile ?? true;
    const json = keepComments ? stringifyJsonc(obj, null, spaces) : JSON.stringify(obj, null, spaces);
    return newlineAtEndOfFile ? json + '\n' : json;
}

// eslint-disable-next-line @typescript-eslint/ban-types
function injectFormatting<T extends Object>(s: T, format: ContentFormat): T {
    (<any>s)[SymbolFormat] = format;
    return s;
}

// eslint-disable-next-line @typescript-eslint/ban-types
function retrieveFormatting<T extends Object>(s: T): ContentFormat | undefined {
    return (<any>s)[SymbolFormat];
}

interface ContentFormat {
    spaces: string | number | undefined;
    newlineAtEndOfFile: boolean;
}

function hasTrailingNewline(content: string): boolean {
    return /\n$/.test(content);
}

function detectIndent(json: string): string | undefined {
    const s = json.match(/^[ \t]+(?=")/m);
    if (!s) return undefined;
    return s[0];
}

function detectFormatting(content: string): ContentFormat {
    return {
        spaces: detectIndent(content),
        newlineAtEndOfFile: hasTrailingNewline(content),
    };
}

export interface ConfigFileReaderWriter extends ConfigReaderWriter {
    readonly uri: Uri;
    _update(fn: ConfigUpdateFn): Promise<void>;
    _read(): Promise<CSpellUserSettings>;
}

abstract class AbstractConfigFileReaderWriter implements ConfigFileReaderWriter {
    constructor(readonly uri: Uri) {}

    async read<K extends keyof CSpellUserSettings>(keys: K[]): Promise<Pick<CSpellUserSettings, K>> {
        return extractKeys(await this._read(), keys);
    }

    abstract _read(): Promise<CSpellUserSettings>;
    abstract write(settings: CSpellSettings): Promise<void>;

    update<K extends keyof CSpellUserSettings>(fn: ConfigUpdateFn, keys: K[]): Promise<void> {
        return this._update((cfg) => fn(extractKeys(cfg, keys)));
    }

    abstract _update(fn: ConfigUpdateFn): Promise<void>;

    mkdir(): Promise<void> {
        return fs.mkdirp(Uri.joinPath(this.uri, '..').fsPath);
    }
}

class ConfigFileReaderWriterJson extends AbstractConfigFileReaderWriter {
    async _update(updateFn: ConfigUpdateFn): Promise<void> {
        const cspell = await orDefault(this._read(), settingsFileTemplate);
        const updated = assignJson(cspell, updateFn(cspell));
        return this.write(updated);
    }

    async write(cfg: CSpellSettings): Promise<void> {
        await this.mkdir();
        const json = stringifyJson(cfg);
        const content = json[json.length - 1] !== '\n' ? json + '\n' : json;
        return fs.writeFile(this.uri.fsPath, content);
    }

    async _read(): Promise<CSpellSettings> {
        const uri = this.uri;
        const content = await fs.readFile(uri.fsPath, 'utf8');
        const s = parseJson(content) as CSpellSettings;
        return s;
    }
}

class ConfigFileReaderWriterPackage extends AbstractConfigFileReaderWriter {
    async _update(updateFn: ConfigUpdateFn): Promise<void> {
        const fsPath = this.uri.fsPath;
        const content = await fs.readFile(fsPath, 'utf8');
        const pkg = parseJson(content) as PackageWithCSpell;
        const cspell = pkg.cspell || { ...settingsFileTemplate };
        pkg.cspell = Object.assign(cspell, updateFn(cspell));
        return fs.writeFile(fsPath, stringifyJson(pkg, spacesPackage, false));
    }

    async write(cfg: CSpellSettings): Promise<void> {
        const fsPath = this.uri.fsPath;
        const content = await fs.readFile(fsPath, 'utf8');
        const pkg = parseJson(content) as PackageWithCSpell;
        pkg.cspell = cfg;
        return fs.writeFile(fsPath, stringifyJson(pkg, spacesPackage, false));
    }

    async _read(): Promise<CSpellSettings> {
        const content = await fs.readFile(this.uri.fsPath, 'utf8');
        const pkg = parseJson(content) as { cspell?: CSpellPackageSettings };
        if (!pkg.cspell || typeof pkg.cspell !== 'object') {
            throw new SysLikeError('`cspell` section missing from package.json', 'ENOENT');
        }
        return pkg.cspell;
    }
}

class ConfigFileReaderWriterYaml extends AbstractConfigFileReaderWriter {
    async _update(updateFn: ConfigUpdateFn): Promise<void> {
        const cspell = await orDefault(this._read(), settingsFileTemplate);
        const updated = Object.assign({}, cspell, updateFn(cspell));
        return this.write(updated);
    }

    async write(cfg: CSpellSettings): Promise<void> {
        await this.mkdir();
        return fs.writeFile(this.uri.fsPath, stringifyYaml(cfg));
    }

    async _read(): Promise<CSpellSettings> {
        const content = await fs.readFile(this.uri.fsPath, 'utf8');
        return parseYaml(content) as CSpellSettings;
    }
}

export const __testing__ = {
    settingsFileTemplate,
    detectIndent,
    injectFormatting,
    SymbolFormat,
};
