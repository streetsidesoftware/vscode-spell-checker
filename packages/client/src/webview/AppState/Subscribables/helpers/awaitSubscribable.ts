import type { DisposableLike } from 'utils-disposables';
import { disposeOf } from 'utils-disposables';

import type { EventType, Subscribable, SubscribableEvent } from '../Subscribables';

export function awaitSubscribable<T>(sub: Subscribable<T>): Promise<T> {
    let disposable: DisposableLike | undefined;

    return new Promise<T>((resolve, _reject) => {
        // prevent double resolve.
        let resolved = false;
        disposable = sub.subscribe((v) => {
            if (resolved) return;
            resolved = true;
            try {
                resolve(v);
            } finally {
                disposeOf(disposable);
            }
        });
    });
}

export function awaitSubscribableAll<T>(sub: Subscribable<T>): Promise<T[]> {
    let disposable: DisposableLike | undefined;

    return new Promise<T[]>((resolve, _reject) => {
        // prevent double resolve.
        let resolved = false;
        const buffer: T[] = [];

        function onDone() {
            if (resolved) return;
            resolved = true;
            disposeOf(disposable);
            resolve(buffer);
        }

        disposable = sub.subscribe({ notify: (v) => buffer.push(v), done: onDone });
    });
}

export function awaitEvent<T>(sub: Subscribable<T>, eventName: EventType): Promise<SubscribableEvent> {
    let disposable: DisposableLike | undefined;

    return new Promise((resolve) => {
        disposable = sub.onEvent((e) => {
            if (e.name == eventName) {
                disposeOf(disposable);
                resolve(e);
            }
        });
    });
}
