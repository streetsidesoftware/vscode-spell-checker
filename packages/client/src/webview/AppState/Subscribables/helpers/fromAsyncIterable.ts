import { disposeOf } from 'utils-disposables';

import { createEmitter } from '../createFunctions';
import type { Subscribable, SubscribableEvent } from '../Subscribables';

export function fromAsyncIterable<T>(iter: AsyncIterableIterator<T> | AsyncIterable<T>): Subscribable<T> {
    const emitter = createEmitter<T>();
    let stop = false;
    const disposeEventListener = emitter.onEvent(handleEvents);

    function handleEvents(e: SubscribableEvent) {
        if (e.name === 'onStart') {
            setTimeout(emitValues);
        }
        if (e.name === 'onStop' || e.name === 'onDone') {
            stop = true;
        }
    }

    async function emitValues() {
        try {
            for await (const val of iter) {
                if (stop) break;
                emitter.notify(val);
            }
        } finally {
            disposeOf(disposeEventListener);
            emitter.done();
        }
    }

    return emitter;
}
