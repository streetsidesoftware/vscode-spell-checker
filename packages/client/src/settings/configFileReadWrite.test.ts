import { fsRemove, getPathToTemp, writeFile, readFile } from '../test/helpers';
import { readConfigFile, updateConfigFile, writeConfigFile, __testing__ } from './configFileReadWrite';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import { parse as parseJsonc, stringify as stringifyJsonc, assign } from 'comment-json';

const samplePackageJson = {
    name: 'sample-package',
    scripts: {},
    dependencies: {},
};

const sampleCSpell = {
    words: ['sample', 'one', 'two'],
};

describe('Validate configFileReadWrite', () => {
    test.each`
        filename                | initialContent
        ${'cspell.json'}        | ${undefined}
        ${'cspell.jsonc'}       | ${undefined}
        ${'cspell.config.yaml'} | ${undefined}
        ${'package.json'}       | ${toJson(samplePackageJson)}
    `('update $filename', async ({ filename, initialContent }) => {
        await fsRemove(getPathToTemp());
        const uri = getPathToTemp(filename);
        if (initialContent) {
            await writeFile(uri, initialContent);
        }
        await expect(readConfigFile(uri)).resolves.toBe(undefined);
        await expect(readConfigFile(uri, sampleCSpell)).resolves.toBe(sampleCSpell);
        await expect(readConfigFile(uri)).resolves.toBe(undefined);

        await updateConfigFile(uri, () => sampleCSpell);
        const r = await readConfigFile(uri);
        expect(r).not.toEqual(sampleCSpell);
        expect(r?.words).toEqual(sampleCSpell.words);
        expect(r?.version).toEqual(__testing__.settingsFileTemplate.version);
        expect(await readFile(uri)).toMatchSnapshot();
    });

    test.each`
        filename                | initialContent
        ${'cspell.json'}        | ${undefined}
        ${'cspell.config.json'} | ${toJson(parseJsonc('{}'))}
        ${'cspell.jsonc'}       | ${undefined}
        ${'cspell.config.yaml'} | ${undefined}
        ${'package.json'}       | ${toJson(samplePackageJson)}
    `('write $filename', async ({ filename, initialContent }) => {
        await fsRemove(getPathToTemp());
        const uri = getPathToTemp(filename);
        if (initialContent) {
            await writeFile(uri, initialContent);
        }

        const sample = sampleJsoncObj();
        await writeConfigFile(uri, sample);
        const r = await readConfigFile(uri);
        delete (<any>r)[__testing__.SymbolFormat];
        expect(r).toEqual(sample);
        expect(await readFile(uri)).toMatchSnapshot();
    });

    test('jsonc with comments', async () => {
        await fsRemove(getPathToTemp());
        const uri = getPathToTemp('cspell.jsonc');
        await writeFile(uri, sampleJsonc());
        const sample = parseJsonc(sampleJsonc());
        const sampleWithSpacesSymbol = __testing__.injectFormatting({ ...sample }, { spaces: '    ', newlineAtEndOfFile: true });
        await expect(readConfigFile(uri)).resolves.toEqual(sampleWithSpacesSymbol);

        await updateConfigFile(uri, () => sampleCSpell);
        const r = await readConfigFile(uri);
        expect(r).not.toEqual(sampleCSpell);
        expect(r?.words).toEqual(sampleCSpell.words);
        expect(r?.version).toEqual(__testing__.settingsFileTemplate.version);

        const configFileContext = await readFile(uri);
        expect(configFileContext).toBe(toJson(assign(sampleJsoncObj(), sampleCSpell)));
    });

    test('yaml with comments', async () => {
        await fsRemove(getPathToTemp());
        const uri = getPathToTemp('cspell.yaml');
        await writeFile(uri, sampleYaml());
        const sample = parseYaml(sampleYaml());
        await expect(readConfigFile(uri)).resolves.toEqual(sample);

        await updateConfigFile(uri, () => ({}));
        const r = await readConfigFile(uri);
        expect(r).not.toEqual(sampleCSpell);
        expect(r?.words).toEqual(sampleCSpell.words);
        expect(r?.version).toEqual(__testing__.settingsFileTemplate.version);

        const configFileContext = await readFile(uri);
        // Note: comments are lost.
        expect(configFileContext).toBe(stringifyYaml(parseYaml(sampleYaml())));
    });

    test.each`
        spaces
        ${2}
        ${1}
        ${4}
        ${'\t'}
        ${'   '}
        ${undefined}
    `('Detect space "$spaces"', ({ spaces }) => {
        const obj = sampleJsoncObj();
        const json = stringifyJsonc(obj, null, spaces);
        const detected = __testing__.detectIndent(json);
        const expected = typeof spaces === 'number' ? ' '.repeat(spaces) : spaces;
        expect(detected).toBe(expected);
    });
});

function toJson(obj: any): string {
    return stringifyJsonc(obj, null, 4) + '\n';
}

function sampleJsoncObj(): any {
    return parseJsonc(sampleJsonc());
}

function sampleJsonc(): string {
    return `
// This is a Json file with comments.
{
    // Version should be 0.2
    "version": "0.2",
    // Custom terms should be here.
    "words": [],
    // Words to be ignored
    "ignoreWords": []
}
`;
}

function sampleYaml(): string {
    return `
# First comment
version: "0.2" # Version should always be 0.2
ignorePaths: []
dictionaryDefinitions: []
dictionaries: []
words:
    - sample
    - one
    - two
ignoreWords: []
import: []
`;
}
