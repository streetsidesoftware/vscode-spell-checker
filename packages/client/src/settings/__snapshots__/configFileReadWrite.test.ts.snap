// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate configFileReadWrite update cspell.config.yaml 1`] = `
"version: \\"0.2\\"
ignorePaths: []
dictionaryDefinitions: []
dictionaries: []
words:
  - sample
  - one
  - two
ignoreWords: []
import: []
"
`;

exports[`Validate configFileReadWrite update cspell.json 1`] = `
"{
    \\"version\\": \\"0.2\\",
    \\"ignorePaths\\": [],
    \\"dictionaryDefinitions\\": [],
    \\"dictionaries\\": [],
    \\"words\\": [
        \\"sample\\",
        \\"one\\",
        \\"two\\"
    ],
    \\"ignoreWords\\": [],
    \\"import\\": []
}
"
`;

exports[`Validate configFileReadWrite update cspell.jsonc 1`] = `
"{
    \\"version\\": \\"0.2\\",
    \\"ignorePaths\\": [],
    \\"dictionaryDefinitions\\": [],
    \\"dictionaries\\": [],
    \\"words\\": [
        \\"sample\\",
        \\"one\\",
        \\"two\\"
    ],
    \\"ignoreWords\\": [],
    \\"import\\": []
}
"
`;

exports[`Validate configFileReadWrite update package.json 1`] = `
"{
    \\"name\\": \\"sample-package\\",
    \\"scripts\\": {},
    \\"dependencies\\": {},
    \\"cspell\\": {
        \\"version\\": \\"0.2\\",
        \\"ignorePaths\\": [],
        \\"dictionaryDefinitions\\": [],
        \\"dictionaries\\": [],
        \\"words\\": [
            \\"sample\\",
            \\"one\\",
            \\"two\\"
        ],
        \\"ignoreWords\\": [],
        \\"import\\": []
    }
}
"
`;

exports[`Validate configFileReadWrite write cspell.config.json 1`] = `
"// This is a Json file with comments.
{
    // Version should be 0.2
    \\"version\\": \\"0.2\\",
    // Custom terms should be here.
    \\"words\\": [],
    // Words to be ignored
    \\"ignoreWords\\": []
}
"
`;

exports[`Validate configFileReadWrite write cspell.config.yaml 1`] = `
"version: \\"0.2\\"
words: []
ignoreWords: []
"
`;

exports[`Validate configFileReadWrite write cspell.json 1`] = `
"// This is a Json file with comments.
{
    // Version should be 0.2
    \\"version\\": \\"0.2\\",
    // Custom terms should be here.
    \\"words\\": [],
    // Words to be ignored
    \\"ignoreWords\\": []
}
"
`;

exports[`Validate configFileReadWrite write cspell.jsonc 1`] = `
"// This is a Json file with comments.
{
    // Version should be 0.2
    \\"version\\": \\"0.2\\",
    // Custom terms should be here.
    \\"words\\": [],
    // Words to be ignored
    \\"ignoreWords\\": []
}
"
`;

exports[`Validate configFileReadWrite write package.json 1`] = `
"{
    \\"name\\": \\"sample-package\\",
    \\"scripts\\": {},
    \\"dependencies\\": {},
    \\"cspell\\": {
        \\"version\\": \\"0.2\\",
        \\"words\\": [],
        \\"ignoreWords\\": []
    }
}
"
`;
